src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,171 @@
-		// TODO Auto-generated method stub
-		return null;
+		// The following implementation of this method is inefficient because,
+		// each time it is called, it creates and populates a list of solution
+		// mappings by iterating over the content of this index. However,
+		// for a thread-safe implementation, we may have to live with this.
+		// TODO: is there a more efficient way to implement this method?
+		final List<SolutionMapping> solMap = new ArrayList<>();
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while(li.hasNext()){
+			solMap.addAll(li.next());
+		}
+		return solMap.iterator();
 	}
 
 	@Override
 	public boolean add( final SolutionMapping e ) {
-		// TODO Auto-generated method stub
-		return false;
+		final List<Node> valKeys = getVarKeys(e);
+
+		if (map.containsKey(valKeys)) {
+			map.get(valKeys).add(e);
+		}
+		else {
+			final List<SolutionMapping> solMapList = new ArrayList<>();
+			solMapList.add(e);
+			map.put(valKeys, solMapList);
+		}
+		return true;
 	}
 
 	@Override
 	public void clear() {
-		// TODO Auto-generated method stub
-		
+		map.clear();
 	}
 
 	@Override
 	public Iterable<SolutionMapping> getJoinPartners( final SolutionMapping sm )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		final List<Node> valKeys = getVarKeys(sm);
+		final List<SolutionMapping> joinPartner = new ArrayList<>();
+		List<SolutionMapping> matchingSolMaps = new ArrayList<>();
+
+		if (!valKeys.isEmpty() && valKeys.size() == joinVariables.size()){
+			matchingSolMaps = map.get(valKeys);
+		}
+		else {
+			final Iterator<List<SolutionMapping>> li = map.values().iterator();
+			while(li.hasNext()){
+				matchingSolMaps.addAll(li.next());
+			}
+		}
+
+		if (matchingSolMaps != null) {
+			for (SolutionMapping matchSolM : matchingSolMaps) {
+				if (SolutionMappingUtils.compatible(sm, matchSolM)) {
+					joinPartner.add(matchSolM);
+				}
+			}
+		}
+		return joinPartner;
 	}
 
 	@Override
 	public Iterable<SolutionMapping> findSolutionMappings( final Var var, final Node value )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		if ( joinVariables.size() == 1 && joinVariables.contains(var) ){
+			final List<Node> valKeyL = new ArrayList<>();
+			valKeyL.add(value);
+			final List<SolutionMapping> solMaps = map.get(valKeyL);
+			return solMaps;
+		}
+		else{
+			throw new IllegalArgumentException();
+		}
 	}
 
 	@Override
 	public Iterable<SolutionMapping> findSolutionMappings(
 			final Var var1, final Node value1,
 			final Var var2, final Node value2 )
-					throws UnsupportedOperationException
+			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		if ( joinVariables.size() == 2 && joinVariables.contains(var1) && joinVariables.contains(var2) ){
+			final List<Node> valKeyL = new ArrayList<>();
+			valKeyL.add(value1);
+			valKeyL.add(value2);
```

Comment:
This may be incorrect. You cannot assume that `var1` and `var2` are given in the order in which these two variables appear in `joinVariables`. Hence, you have to check whether `var1` or `var2` is the first variable in `joinVariables` and, depending on this check, add `value1` first or add `value2` first.